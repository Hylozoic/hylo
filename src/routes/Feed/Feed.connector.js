import { connect } from 'react-redux'
import { push } from 'connected-react-router'
import { get } from 'lodash/fp'
import { isEmpty } from 'lodash'
import {
  FETCH_POSTS, FETCH_FOR_CURRENT_USER,
  FETCH_TOPIC, FETCH_GROUP_TOPIC
} from 'store/constants'
import getGroupForCurrentRoute from 'store/selectors/getGroupForCurrentRoute'
import getGroupTopicForCurrentRoute from 'store/selectors/getGroupTopicForCurrentRoute'
import getTopicForCurrentRoute from 'store/selectors/getTopicForCurrentRoute'
import getRouteParam from 'store/selectors/getRouteParam'
import getMe from 'store/selectors/getMe'
import getMyMemberships from 'store/selectors/getMyMemberships'
import getQuerystringParam from 'store/selectors/getQuerystringParam'
import changeQuerystringParam from 'store/actions/changeQuerystringParam'
import toggleGroupTopicSubscribe from 'store/actions/toggleGroupTopicSubscribe'
import { createGroupUrl, createPostUrl } from 'util/navigation'
import { fetchTopic, fetchGroupTopic } from './Feed.store'
import isPendingFor from 'store/selectors/isPendingFor'

export function mapStateToProps (state, props) {
  let group, groupTopic, topic

  const routeParams = get('match.params', props)
  const location = get('location', props)
  const view = routeParams.view
  const querystringParams = getQuerystringParam(['s', 't'], null, props)
  const currentUser = getMe(state)
  const currentUserHasMemberships = !isEmpty(getMyMemberships(state))
  const groupSlug = getRouteParam('groupSlug', state, props)
  const topicName = getRouteParam('topicName', state, props)
  const topicLoading = isPendingFor([FETCH_TOPIC, FETCH_GROUP_TOPIC], state)

  if (groupSlug) {
    group = getGroupForCurrentRoute(state, props)
    groupTopic = getGroupTopicForCurrentRoute(state, props)
    groupTopic = groupTopic && { ...groupTopic.ref, group: groupTopic.group, topic: groupTopic.topic }
  }
  if (topicName) {
    topic = getTopicForCurrentRoute(state, props)
  }

  const postTypeFilter = view === 'projects' ? 'project' : (view === 'events' ? 'event' : getQuerystringParam('t', state, props))
  const sortBy = getQuerystringParam('s', state, props)

  return {
    routeParams,
    querystringParams,
    location,
    postTypeFilter,
    sortBy,
    currentUser,
    currentUserHasMemberships,
    groupTopic,
    groupSlug,
    group,
    topicLoading,
    topicName,
    topic,
    postsTotal: get('postsTotal', groupSlug ? groupTopic : topic),
    followersTotal: get('followersTotal', groupSlug ? groupTopic : topic),
    selectedPostId: getRouteParam('postId', state, props),
    membershipsPending: state.pending[FETCH_FOR_CURRENT_USER],
    postCount: get('postCount', group),
    pending: state.pending[FETCH_POSTS]
  }
}

export function mapDispatchToProps (dispatch, props) {
  const groupSlug = getRouteParam('groupSlug', null, props)
  const topicName = getRouteParam('topicName', null, props)
  const routeParams = get('match.params', props)
  const querystringParams = getQuerystringParam(['s', 't'], null, props)

  return {
    changeTab: tab => dispatch(changeQuerystringParam(props, 't', tab, 'all')),
    changeSort: sort => dispatch(changeQuerystringParam(props, 's', sort, 'all')),
    fetchTopic: () => {
      if (groupSlug && topicName) {
        return dispatch(fetchGroupTopic(topicName, groupSlug))
      } else if (topicName) {
        return dispatch(fetchTopic(topicName))
      }
    },
    toggleGroupTopicSubscribe: groupTopic =>
      dispatch(toggleGroupTopicSubscribe(groupTopic)),
    goToCreateGroup: () => dispatch(push(createGroupUrl(routeParams))),
    newPost: () => dispatch(push(createPostUrl(routeParams, querystringParams)))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
