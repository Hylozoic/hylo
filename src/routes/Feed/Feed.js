import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { get, pick } from 'lodash/fp'

import './Feed.scss'
import FeedList from 'components/FeedList'
import Loading from 'components/Loading'
import FeedBanner from 'components/FeedBanner'
import TopicFeedHeader from 'components/TopicFeedHeader'
import Button from 'components/Button'
import { bgImageStyle } from 'util/index'

export default class Feed extends Component {
  static propTypes = {
    newPost: PropTypes.func,
    routeParams: PropTypes.object,
    querystringParams: PropTypes.object
  }

  static defaultProps = {
    routeParams: {},
    querystringParams: {}
  }

  componentDidMount () {
    const { routeParams, fetchTopic } = this.props
    const { topicName } = routeParams

    if (topicName) fetchTopic()
  }

  componentDidUpdate (prevProps) {
    const { routeParams, fetchTopic } = this.props
    const { groupSlug, topicName } = routeParams
    const topicChanged = topicName && get('routeParams.topicName', prevProps) !== topicName
    const slugChanged = groupSlug && get('routeParams.groupSlug', prevProps) !== groupSlug
    if (topicChanged || (topicName && slugChanged)) fetchTopic()
  }

  getFeedProps () {
    const { routeParams, querystringParams } = this.props
    const { context } = routeParams

    return {
      context,
      routeParams,
      querystringParams,
      topic: get('id', this.props.topic),
      groupId: get('group.id', this.props),
      ...pick([
        'postTypeFilter',
        'sortBy',
        'changeSort',
        'changeTab',
        'selectedPostId'
      ], this.props)
    }
  }

  render () {
    const {
      routeParams, topic, group, currentUser, postsTotal, followersTotal,
      groupTopic, newPost, currentUserHasMemberships,
      goToCreateGroup, membershipsPending, postTypeFilter
    } = this.props
    const { topicName, context } = routeParams

    if (topicName && !topic) return <Loading />
    if (group && topicName && !groupTopic) return <Loading />
    if (!currentUser) return <Loading />
    if (membershipsPending) return <Loading />

    return <div>
      {topicName
        ? <TopicFeedHeader
          groupTopic={groupTopic}
          topicName={topicName}
          postsTotal={postsTotal}
          followersTotal={followersTotal}
          topic={topic}
          type={postTypeFilter}
          group={group}
          currentUser={currentUser}
          newPost={newPost} />
        : <FeedBanner
          group={group}
          currentUser={currentUser}
          type={postTypeFilter}
          all={context && context === 'all'}
          publicContext={context && context === 'public'}
          newPost={newPost}
          currentUserHasMemberships={currentUserHasMemberships} />}
      {currentUserHasMemberships && <FeedList {...this.getFeedProps()} />}
      {!membershipsPending && !currentUserHasMemberships && <CreateGroupPrompt
        goToCreateGroup={goToCreateGroup}
      />}
      {membershipsPending && <Loading />}
    </div>
  }
}

export function CreateGroupPrompt ({ goToCreateGroup }) {
  return <div styleName='create-group-prompt'>
    <p>There's no posts yet, try starting a group!</p>
    <Button
      styleName='button'
      label='Create a Group'
      onClick={goToCreateGroup}
    />
    <div style={bgImageStyle('/assets/hey-axolotl.png')} styleName='sidebar-image' />
  </div>
}
