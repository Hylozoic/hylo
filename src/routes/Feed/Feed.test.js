import React from 'react'
import { shallow } from 'enzyme'
import Feed, { CreateGroupPrompt } from './Feed'

describe('Feed', () => {
  it('renders a FeedList with correct props', () => {
    const wrapper = shallow(<Feed
      routeParams={{ slug: 'foo' }}
      postTypeFilter='request'
      group={{}}
      selectedPostId='5'
      sortBy='votes'
      currentUser
      currentUserHasMemberships />)

    const feed = wrapper.find('Connect(Component)')
    expect(feed).toHaveLength(1)
    expect(feed.props()).toEqual({
      routeParams: {
        slug: 'foo'
      },
      querystringParams: {},
      subject: 'group',
      postTypeFilter: 'request',
      sortBy: 'votes',
      selectedPostId: '5',
      topic: undefined
    })
  })

  it('displays the regular FeedBanner if on the main feed', () => {
    const props = {
      currentUser: { id: 1 }
    }
    const wrapper = shallow(<Feed {...props} />)
    expect(wrapper.find('FeedBanner')).toHaveLength(1)
    expect(wrapper.find('Connect(TopicFeedHeader)')).toHaveLength(0)
  })

  it('displays the TopicFeedHeader if on a topic feed', () => {
    const props = {
      routeParams: {
        topicName: 'petitions'
      },
      topic: { id: '5', name: 'petitions' },
      currentUser: { id: 1 }
    }
    const wrapper = shallow(<Feed {...props} fetchTopic={jest.fn()} />)
    expect(wrapper.find('FeedBanner')).toHaveLength(0)
    expect(wrapper.find('Connect(TopicFeedHeader)')).toHaveLength(1)
  })
})

describe('CreateGroupPrompt', () => {
  it('matches the latest snapshot', () => {
    const wrapper = shallow(<CreateGroupPrompt
      goToCreateGroup={jest.fn()}
    />)
    expect(wrapper).toMatchSnapshot()
  })
})
