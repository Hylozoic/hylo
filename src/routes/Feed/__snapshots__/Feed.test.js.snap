// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateGroupPrompt matches the latest snapshot 1`] = `
<div
  data-stylename="create-group-prompt"
>
  <p>
    There's no posts yet, try starting a group!
  </p>
  <Button
    data-stylename="button"
    label="Create a Group"
    onClick={[MockFunction]}
  />
  <div
    data-stylename="sidebar-image"
    style={
      Object {
        "backgroundImage": "url(/assets/hey-axolotl.png)",
      }
    }
  />
</div>
`;

exports[`Feed with a network it renders a Feed with correct props 1`] = `
<div>
  <FeedBanner
    group={
      Object {
        "id": 2,
      }
    }
    currentUser={true}
    currentUserHasMemberships={true}
    type="request"
  />
  <Connect(Component)
    postTypeFilter="request"
    querystringParams={Object {}}
    routeParams={
      Object {
        "networkSlug": "bar",
      }
    }
    sortBy="votes"
    subject="network"
  />
</div>
`;
